fenwick/src/main/java/org/opencadc/fenwick/InventoryHarvester.java

```diff
@@ -197,9 +224,35 @@ public void run() {
                 // ... this value and the reprocess-last-N-seconds should be related
                 long dt = 60 * 1000L;
                 Thread.sleep(dt);
+
+                // successful run, reset retry values
+                retry = false;
+                retryCount = 1;
+                sleepSeconds = INITIAL_SLEEP_TIMEOUT;
             } catch (PrivilegedActionException privilegedActionException) {
-                final Exception exception = privilegedActionException.getException();
-                throw new IllegalStateException(exception.getMessage(), exception);
+                // Get the check exceptions in PrivilegedActionException thrown by doit()
+                final Exception cause = privilegedActionException.getException();
```

Comment:
Ugh. This PrivilegedActionException thing is a pain because the resulting conditional is really error prone.

First, `doit()` should never be throwing NoSuchAlgorithmException: track down where that comes from, catch it as locally as possible, and wrap it in a `new RuntimeException("BUG: failed to get instance of ?", ..., ex)`.

Second issue you did not seem to recognise: we use RuntimeException for local bugs and unexpected failures (as above) all over the place, and the cadc-rest library translates a RuntimeException into a 500, so the HttpTransfer (client and subclasses) translates a 500 back into a RuntimeException... that's seems fine except you cannot tell the difference between a local and remote RuntimeException. So, the first thing to do is probably go into cadc-util and add a new Exception: `RemoteServiceException extends RuntimeException` and use that for a 500 response in `HttpTransfer.checkErrors`. Without that you will not be able to deal with the important case (some service responding with a  500 because it's broken).

And then there is this PrivilegedExceptionAction bit that completely messes up exception handling. The current code organisation means you can't use catch (multi-catch) to handle exceptions so I think you have to rethink where the Subject.doAs(...) lives. It's in the current spot because it was the easy way to re-read the proxy cert file before each query rather... I think all this error handling has to be inside there so you can use normal exception handling. Remember that Subject.doAs(...) can return a value (the `PrivilegedExceptionAction<Void>` just says it doesn't, but the type arg could be Boolean or String or whatever class you need to convey info from inside... or you could just handle all retry-able exceptions inside and any throw is fatal. 

Side note: it took me quite a bit of work to modify the ca.nrc.cadc.net package to have a prepare() method that did all the work (except the transfer) and throw declared exceptions in the normal way, but it is a comparable thing. The old style of `run()` and then `getThrowable()` was fine if you were going to just print the message, but horrible if you wanted to actually write code to react to different failures. The use or misplacing of a Subject.doAs(...) wrecks havoc on error handling.... so this will take some thought to come up with a clean and maintainable solution.

Reply:
Removed the throws declarations for NoSuchAlgorithmException, now catch NoSuchAlgorithmException where it is thrown and then throw a RuntimeException.

Added a RemoteServiceException class to the cadc-util net classes, updated HttpTransfer to throw it for a HTTP 500 error.

Updated the Subject.doAs to use a PrivilegedAction and catch the checked exceptions inside the run() method. The action returns a boolean indicating whether an exception was thrown that should result in a retry.

Fatal runtime exceptions are caught in the try around the Subject.doAs.

You mentioned moving the Subject.doAs out of the loop. You could put the .doAs into a method and call the method, but I think it's better to keep all the exception handling and retry/fail logic in the same place. Easier to see and understand what it is doing. Open to being wrong though :-)